src/nm-checkpoint.c

```diff
@@ -245,6 +248,15 @@ restore_and_activate_connection (NMCheckpoint *self,
 			g_clear_error (&local_error);
 			return FALSE;
 		}
+
+		/* If the device is software, a brand new NMDevice may have been created */
+		if (   dev_checkpoint->is_software
+		    && !dev_checkpoint->device) {
+			dev_checkpoint->device = nm_manager_get_device (priv->manager,
+			                                                dev_checkpoint->original_dev_name,
+			                                                dev_checkpoint->dev_type);
+			nm_g_object_ref (dev_checkpoint->device);
```

Comment:
Now if no device exits, you won't get an assertion failure during `g_object_ref()`.

But a few lines below, you'll pass `%NULL` to `nm_device_get_state (dev_checkpoint->device)`. I meant, %NULL must be handled, and that *may* be as simple as using `nm_g_object_ref()`. But looking at it, it requires a bit more...

Reply:
Right. Pushed a fixup commit on top of the branch to fix this.

Reply:
there are still several other code paths that lead to
```
»···»···if (   nm_device_get_state (dev_checkpoint->device) > NM_DEVICE_STATE_DISCONNECTED
»···»···    && nm_device_get_state (dev_checkpoint->device) < NM_DEVICE_STATE_DEACTIVATING) {
```

you need to handle the possibility of device being %NULL there...

Reply:
Ok, the point is that we have to account a software device NOT created by NM that is removed after checkpoint creation. This is what is missing in the logic here.
Updated the fixup commit.